2020-05/robot-name/robot-name.js

```diff
@@ -2,46 +2,121 @@
 // provided as a convenience to get your started writing code faster.
 
 export class Robot {
-  constructor(name){
-    this.nameHistory = []
-    this.name = this.randomName()
+  // A list of all robot names used.
+  static nameHistory = {}
+
+  // A private robot name.
+  #name = ''
+
+  constructor() {
+    this.#name = this.newUniqueName()
+  }
+
+  /**
+   * A getter to return the "name" of the current robot.
+   *
+   * @return {string|*}
+   */
+  get name() {
+    return this.#name
   }
 
+  /**
+   * Create the random name for the robot in the format of two uppercase letters
+   * followed by three digits, such as RX837 or BC811.
+   *
+   * @return {string}
+   */
   randomName() {
-    let result = ''
-    result += this.makeid(2)
+    let name = ''
 
-    for (var i = 0; i < 3; i++) {
-      result += Math.floor(Math.random() * 10).toString()
+    // Generate the first two alpha-numeric characters.
+    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+    for (let i = 0; i < 2; i++) {
+      name += characters.charAt(Math.floor(Math.random() * characters.length))
     }
-    // console.log("before adding")
-    // console.log(this.nameHistory)
-    this.nameHistory.push(result)
-    // console.log("after adding")
-    // console.log(this.nameHistory)
-    return result
-  }
-
-  makeid(length) {
-    var result = '';
-    var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
-    var charactersLength = characters.length;
-    for ( var i = 0; i < length; i++ ) {
-    result += characters.charAt(Math.floor(Math.random() * charactersLength));
+
+    // Generate the remaining three random numbers.
+    for (let i = 0; i < 3; i++) {
+      name += Math.floor(Math.random() * 10).toString()
     }
-    return result;
+
+    return name
+  }
+
+  /**
+   * Generate a new Unique name for a robot.
+   *
+   * @return {string}
+   */
+  newUniqueName() {
+    let name = ''
+    do {
+      name = this.randomName()
+    } while (this.notUnique(name))
+
+    this.addToHistory(name)
+    return name
+  }
+
+  /**
+   * Check if a name is not unique.
+   *
+   * @param name
+   * @return {boolean|*}
+   */
+  notUnique(name) {
+    const { alpha, numeric } = this.splitName(name)
+
+    if (Robot.nameHistory[alpha] !== undefined) {
+      return Robot.nameHistory[alpha].includes(numeric)
+    } else {
+      return false
     }
+  }
 
-  reset() {
+  /**
+   * Add the name into the list of used names.
+   *
+   * @param name
+   */
+  addToHistory(name) {
+    const { alpha, numeric } = this.splitName(name)
 
-    let newName = this.randomName()
-    if (this.nameHistory.includes(newName)) {
-      this.reset()
+    if (Robot.nameHistory[alpha] !== undefined) {
+      Robot.nameHistory[alpha].push(numeric)
+    } else {
+      Robot.nameHistory[alpha] = [numeric]
     }
```

Comment:
Restating (for others since I know Mark knows this)

This function is storing the used names in a smarter way. Instead of each name being its own item in an Array, it splits the name into its `alpha` part (e.g. for `XY123`, `alpha` = `XY`) and it's `numeric` part (e.g. for `XY123`, `numeric` = `123`). Then it stores the name in a Key:Value object, with the `alpha` as the key, and the value being a list of `numeric`s. (So `XY123` would be stored as `{XY: [123]}`. Then if we get `XY456`, it would be stored as another numeric in that same `XY` list (`{XY: [123, 456]}`

This is cool, because you can cut down the search space significantly when your checking for a name, because you do a lookup with the `alpha` part of your new name, and then search through the list of numeric parts that go with it. There 676 unique `alpha` strings, so this cuts down the search from 676000 unique names, to 1000 numerics.

Very nice optimization!